edgePy/DGEList.py

```diff
@@ -272,37 +306,120 @@ def library_size(self) -> np.array:
         """
         return np.sum(self.counts, 0)
 
-    def cpm(self, log: bool = False, prior_count: float = PRIOR_COUNT) -> None:
-        """Return the DGEList normalized to read counts per million."""
-        self.counts = 1e6 * self.counts / np.sum(self.counts, axis=0)
-        if log:
-            self.counts[self.counts == 0] = prior_count
-            self.counts = np.log(self.counts)
+    def log_transform(self, counts, prior_count):
+        """Compute the log of the counts"""
+        counts[counts == 0] = prior_count
+        return np.log(counts)
+
+    def cpm(self, transform_to_log: bool = False, prior_count: float = PRIOR_COUNT) -> "DGEList":
+        """Normalize the DGEList to read counts per million."""
+        counts = 1e6 * self.counts / np.sum(self.counts, axis=0)
+        current_log = self.log
+        if transform_to_log:
+            counts = self.log_transform(counts, prior_count)
+            current_log = True
+            
+        return self.copy(counts=counts, current_log=current_log)    
 
     def rpkm(
-        self, gene_lengths: Mapping, log: bool = False, prior_count: float = PRIOR_COUNT
-    ) -> None:
+        self, gene_data: ImportCanonicalData, log: bool = False, prior_count: float = PRIOR_COUNT
+    ) -> "DGEList":
         """Return the DGEList normalized to reads per kilobase of gene length
-        per million reads.
+        per million reads. (RPKM =   numReads / ( geneLength/1000 * totalNumReads/1,000,000 )
 
         """
-        raise NotImplementedError
-
-        # TODO: Implement here
-        # self = self.cpm(log=log, prior_count=prior_count)
+        current_log = self.log
+        temp_gene_len = []
+
+        if self.log:
+            self.counts = np.exp(self.counts)
+            current_log = False
+
+        gene_mask = []
+        gene_ensg = []
+        col_sum = np.sum(self.counts, axis=0)
+
+        for gene in self.genes:
+            if gene.startswith("ENSG"):
+                gene_name = gene
+                gene_ensg.append(gene_name)
+                if gene_data.has_gene(gene_name):
+                    gene_mask.append(True)
+                    temp_gene_len.append(
+                        gene_data.get_length_of_canonical_transcript(gene_name) / 1e3
+                    )
+                else:
+                    gene_mask.append(False)
```

Comment:
processing of the gene_data should go into a separate (probably private function). This way, the rpkm function will be more readable.
Furthermore I would suggest to keep the rpkm more low level (compare the tpm function where I used a generic numpy array which can be anything). This way, users are independent and can use their own transcript lengths.

While most users still simply want to use the canonical transcript length (so nice you coded this!) I would suggest to provide this chunk of code in a utility function so that it can be used like this

```from edgePy.utils import compute_canonical_transcript_lengths
genes=['geneA', 'geneB', ...]
dge = DGEList(..., genes=genes)
gene_lengths = compute_canonical_transcript_lengths(genes)
dge.rpkm(gene_lengths=gene_lengths)
```